### DISCUSSION TITLE
[ORACLE] Cannot get distinct results when using limit + a type decorator (column_expression)

### DISCUSSION THREAD

**Discussion by NicolasM-Forsk** [sqlalchemy/sqlalchemy#8219] [Usage Questions]
### Describe the bug

Hello,

I obtain different results when applying the same custom type and query to Oracle and SQLite.
I want to interpret as boolean a column that can contain 0 (false), any other value (true), or NULL (None).
It works fine until I query a table with both distinct and limit: column_expression is applied sooner with Oracle than with SQLite.
This query returns 1 result with SQLite (expected behavior) and 2 results with Oracle.

```
class Client(Base):
    __tablename__ = 'client'
    id = Column(Integer, primary_key=True)
    active = Column(CustomBoolean)

q = session.query(Client) \
    .with_entities(Client.active) \
    .filter_by(active=True)\
    .limit(200) \
    .distinct()
```

Oracle generated query:
```
SELECT CASE WHEN (CASE WHEN (anon_1.client_active != 0) THEN 1 ELSE anon_1.client_active END != 0) THEN 1 ELSE CASE WHEN (anon_1.client_active != 0) THEN 1 ELSE anon_1.client_active END END AS client_active 
FROM (**SELECT DISTINCT** client.active AS client_active 
FROM client 
WHERE CASE WHEN (client.active != 0) THEN 1 ELSE client.active END = 1) anon_1 
WHERE ROWNUM <= [POSTCOMPILE_param_1]
```

SQLite generated query:
```
**SELECT DISTINCT** CASE WHEN (CASE WHEN (client.active != 0) THEN 1 ELSE client.active END != 0) THEN 1 ELSE CASE WHEN (client.active != 0) THEN 1 ELSE client.active END END AS client_active 
FROM client 
WHERE CASE WHEN (client.active != 0) THEN 1 ELSE client.active END = 1
 LIMIT ? OFFSET ?
```

### To Reproduce

```python
from sqlalchemy import Boolean, Integer, String, Table, Column, MetaData, create_engine, inspect, select
import sqlalchemy
from sqlalchemy.orm import sessionmaker
from sqlalchemy.ext.declarative import declarative_base

import sqlite3
from pathlib import Path

PROVIDER = "oracle"

if PROVIDER == "oracle":
    # CREATE TABLE client
    # (
    #    "ID" NUMBER(10,0),
    #    "ACTIVE" NUMBER(1,0)
    # );
    pass
else:
    Path("tmp.sqlite").unlink(missing_ok=True)
    lite = sqlite3.connect("tmp.sqlite")
    lite.execute("""create table client(
        id INTEGER PRIMARY KEY,
        active INTEGER
    )""")
    lite.execute("""insert into client(active) values(1), (-1), (0), (NULL)""")
    lite.commit()
    lite.close()


class CustomBoolean(sqlalchemy.types.TypeDecorator):
    """0 = False
    -1 or 1 = True
    NULL = None
    By default, SQLAlchemy only supports 0 and 1"""

    impl = Boolean
    cache_ok = True

    # In select or group by clause, replace col with expression: (CASE WHEN col != 0 THEN 1 ELSE col end)
    def column_expression(self, col):
        expr = sqlalchemy.case(
            (col != sqlalchemy.literal_column("0", type_=Integer), sqlalchemy.literal_column("1", type_=Integer)),
            else_=col.expr,
        )
        expr.type = Boolean()
        return expr

    # In where clause, replace col with expression: (CASE WHEN col != 0 THEN 1 ELSE col end)
    class comparator_factory(Boolean.Comparator):
        def normalise(self):
            return sqlalchemy.case(
                (self != sqlalchemy.literal_column("0", type_=Integer), sqlalchemy.literal_column("1", type_=Integer)),
                else_=self.expr
            )

        def __init__(self, expr):
            self.expr = expr
            self.type = expr.type
            self.expr = self.normalise()


Base = declarative_base()


class Client(Base):
    __tablename__ = 'client'
    id = Column(Integer, primary_key=True)
    active = Column(CustomBoolean)


if PROVIDER == "oracle":
    engine = create_engine('...')
else:
    engine = create_engine('sqlite:///tmp.sqlite')
    Base.metadata.create_all(engine)


Session = sessionmaker(bind=engine)
session = Session()

# KO

q = session.query(Client) \
    .with_entities(Client.active) \
    .filter_by(active=True)\
    .limit(200) \
    .distinct()

print(q)
for c in q.all():
    print(c)
print()

# OK

subq = select(Client.id) \
    .filter_by(active=True)\
    .limit(200)

mainq = session.query(Client) \
    .with_entities(Client.active) \
    .filter(Client.id.in_(subq)) \
    .distinct()

print(mainq)
for c in mainq.all():
    print(c)
```

### Error

Distinct clause is not honored with Oracle: query returns `[True, True]`

### Versions

- OS: Ubuntu 20.04
- Python: 3.8.10
- SQLAlchemy: 1.4.22
- Database: 
- cx-Oracle==8.2.1

**Comment by zzzeek** [comment]
hi -

there's no bug in SQLAlchemy demonstrated here.   overall I would say the approach of CustomBoolean is too complicated if you want to also use SQLAlchemy's "LIMIT" emulation for Oracle.  Oracle has no LIMIT clause, hence the Oracle dialect performs a complex translation into a SELECT + subquery that's probably not very efficient in any case in order to simulate LIMIT.

I dont have any test data to test with here but it looks like the column_expression() part is likely what's causing problems, I dont think its necessary as this is just interpreting integrers as booleans.  Do it client side:

```py
class CustomBoolean(TypeDecorator):
    """0 = False
    -1 or 1 = True
    NULL = None
    By default, SQLAlchemy only supports 0 and 1"""

    impl = Boolean
    cache_ok = True


    def process_result_value(self, value, dialect):
        if value is not None:
            value = bool(value != 0)
        return value

    # In where clause, replace col with expression: (CASE WHEN col != 0 THEN 1 ELSE col end)
    class comparator_factory(Boolean.Comparator):
        def normalise(self):
            return case(
                (self != literal_column("0", type_=Integer), literal_column("1", type_=Integer)),
                else_=self.expr
            )

        def __init__(self, expr):
            self.expr = expr
            self.type = expr.type
            self.expr = self.normalise()
```

query here looks like:

```sql
SELECT anon_1.client_active 
FROM (SELECT DISTINCT client.active AS client_active 
FROM client 
WHERE CASE WHEN (client.active != 0) THEN 1 ELSE client.active END = 1) anon_1 
WHERE ROWNUM <= 200
```

**Comment by NicolasM-Forsk** [comment]
There were indeed recursivity problems with my implementation, here is a better one:
```
class CustomBoolean(sqlalchemy.types.TypeDecorator):
    """0 = False
    -1 or 1 = True
    NULL = None
    By default, SQLAlchemy only supports 0 and 1"""

    impl = Boolean
    cache_ok = True

    # In select or group by clause, replace col with expression: (CASE WHEN col != 0 THEN 1 ELSE col end)
    def column_expression(self, col):
        expr = col.expr
        if isinstance(expr.type, CustomBoolean):
            res = sqlalchemy.case(
                (type_coerce(expr, Integer()) != sqlalchemy.literal_column("0", type_=Integer), sqlalchemy.literal_column("1", type_=Integer)),
                else_=type_coerce(expr, Integer())
            )
            return type_coerce(res, Boolean())

        expr.type = Boolean()
        return expr

    # In where clause, replace col with expression: (CASE WHEN col != 0 THEN 1 ELSE col end)
    class comparator_factory(Boolean.Comparator):
        def operate(self, op, other, **kwargs):
            op1 = self
            if hasattr(op1, "__clause_element__"):
                clause = op1.__clause_element__()
                if isinstance(clause.type, CustomBoolean):
                    op1 = self.normalise(clause)

            op2 = other
            if hasattr(op2, "__clause_element__"):
                clause = op2.__clause_element__()
                if isinstance(clause.type, CustomBoolean):
                    op2 = self.normalise(clause)

            return op(
                op1, op2, **kwargs
            )

        @staticmethod
        def normalise(expr):
            res = sqlalchemy.case(
                (type_coerce(expr, Integer()) != sqlalchemy.literal_column("0", type_=Integer), sqlalchemy.literal_column("1", type_=Integer)),
                else_=type_coerce(expr, Integer())
            )
            return type_coerce(res, Boolean())
```

Oracle:
```
SELECT CASE WHEN (anon_1.client_active != 0) THEN 1 ELSE anon_1.client_active END AS client_active 
FROM (SELECT DISTINCT client.active AS client_active 
FROM client 
WHERE CASE WHEN (client.active != 0) THEN 1 ELSE client.active END = 1) anon_1 
WHERE ROWNUM <= [POSTCOMPILE_param_1]
```

SQLite:
```
SELECT DISTINCT CASE WHEN (client.active != 0) THEN 1 ELSE client.active END AS client_active 
FROM client 
WHERE CASE WHEN (client.active != 0) THEN 1 ELSE client.active END = 1
 LIMIT ? OFFSET ?
```

I have written test data at the top of script.
Process_value is not an option for me: it would not be reliable enough for running arbitrary queries and subqueries. Neither does it make `distinct` work.

But I understand having limit & distinct & column_expression & Oracle is an edge case which may never have been supported.

**Reply by zzzeek** [reply]
I would likely map this column directly as Numeric and then use a [hybrid property](https://docs.sqlalchemy.org/en/14/orm/extensions/hybrid.html) for both the in-python and SQL CASE expression.   the hybrid occurs more as though you wrote the expressions manually and won't be deeply embedded in the statement's type, and likely would be less prone to confusion when things like Oracle LIMIT emulation occurs (also I would not use LIMIT at all, even if this weren't Oracle).

**Reply by NicolasM-Forsk** [reply]
I'm not sure, but don't hybrid properties use other declared properties ?
I don't think it fits my use case (multi-tenant app where only Oracle has this kind of boolean column storage. every boolean column in my tables are handled this way).
If this use case is not supported, I'll probably have to fix data storage.

The correct Oracle query would be: 
```
SELECT anon_1.client_active AS client_active 
FROM (
	SELECT DISTINCT CASE WHEN (client.active != 0) THEN 1 ELSE client.active END AS client_active  
	FROM client
	WHERE CASE WHEN (client.active != 0) THEN 1 ELSE client.active END = 1) anon_1 
WHERE ROWNUM <= 200;
```

**Reply by zzzeek** [reply]
when the Oracle dialect rewrites the query to embed itself inside of a subquery, it has no means of disabling "column_expression()" from being called at the outer level.   

That's the main thing you want to avoid.  if you are writing an Oracle application, at the very least, write out LIMIT using Oracle constructs directly, i.e. make a subquery with ROWNUM or use other means.   SQLAlchemy's LIMIT emulation is a band-aid.

**Reply by zzzeek** [reply]
now what I need to review is why is this not using FETCH NEXT, because we support this, and I can remember coding it for Oracle, so why isn't it there.  let's look

**Reply by zzzeek** [reply]
wow.  well this is unfortunate, it's implemented for MS SQL Server but not Oracle.      that would also fix this problem because it doesnt need a subquery.

**Reply by zzzeek** [reply]
I added #8221 for this

**Reply by NicolasM-Forsk** [reply]
Yup, this is interesting. `Fetch` would simplify all select+limit queries.
Hope to see this implemented soon.

**Reply by zzzeek** [reply]
see next answer, this API is supported now you jut have to call it directly.   i had forgotten.  pushed up a doc now

**Reply by NicolasM-Forsk** [reply]
This works for the "select" api, but can we use it through the "query" api?

**Reply by zzzeek** [reply]
no, just convert queries as needed to modern syntax.  Query is legacy

**Reply by zzzeek** [reply]
this is new code you're writing in any case

**Reply by CaselIT** [reply]
> now what I need to review is why is this not using FETCH NEXT, because we support this, and I can remember coding it for Oracle, so why isn't it there. let's look

The logic was that for oracle people wanted the current version so that was left as the default, unless fetch was used explicitly

**Reply by zzzeek** [reply]
ah the wonderful memory skills of the younger-than-me.    Is something wrong with FETCH?  I'm all set to just make it the default

**Reply by CaselIT** [reply]
I don't think so. let's make it the default for v2 and see if anyone complains

**Reply by zzzeek** [reply]
we can alwys have a param to use the old way

**Comment by zzzeek** [comment]
new answer -

with Oracle, use FETCH:


```py
stmt = (
    select(Client.active).select_from(Client)
    .filter_by(active=True)\
    .fetch(200)
    .distinct()
)

for c in session.execute(stmt).all():
    print(c)

```

SQL:

```sql
SELECT DISTINCT CASE WHEN (CASE WHEN (client.active != 0) THEN 1 ELSE client.active END != 0) THEN 1 ELSE CASE WHEN (client.active != 0) THEN 1 ELSE client.active END END AS active 
FROM client 
WHERE CASE WHEN (client.active != 0) THEN 1 ELSE client.active END = 1
 FETCH FIRST :param_1 ROWS ONLY
```

**Comment by CaselIT** [comment]
Hi @NicolasM-Forsk 

I'm not sure why you need that `CustomBoolean` at all. Oracle works without any problem with the sqlalchemy boolean. see example below:
```py
import sqlalchemy as sa

m = sa.MetaData()
t = sa.Table("t", m, sa.Column("col", sa.Boolean()))

e = sa.create_engine(
    "oracle+cx_oracle://scott:tiger@127.0.0.1:1521/?service_name=XEPDB1",
    echo=True,
)
m.create_all(e)

with e.begin() as c:
    c.execute(
        t.insert(),
        [{"col": True}] * 20 + [{"col": False}] * 20 + [{"col": None}] * 20,
    )
with e.connect() as c:
    print(c.scalars(t.select().distinct()).all())
    print(
        c.scalars(t.select().distinct().limit(10).where(t.c.col == True)).all()
    )
```

**Reply by nmoreaud** [reply]
True, SqlAchemy provides coherent boolean management. However, the databases I work with are filled by other tools which put sometimes -1 and sometimes 1 for true values. 

"Where True" condition in sqlalchemy is replaced with "where col = 1" in queries, which is not valid for my dataset. Group by is also wrong, etc.

**Reply by CaselIT** [reply]
> other tools which put sometimes -1 and sometimes 1 for true values.

Understood, that's unfortunate :(

maybe you could add a virtual column in oracle so that you would let oracle handle that?
something like `active_norm number(1,0) generated always as (active != 0) virtual` (not sure if oracle lets you do just `active != 0`, you may need to use the case expression there)

that could also be generated by sqlalchemy using `Computed` (but I'm guessing the schema is not managed with sqlalchemy)

**Comment by NicolasM-Forsk** [comment]
There is something else I'm wondering: I was planning to use the same models for MSSQL, Oracle, etc.
I would like to declare a column like this (pseudo code) : 
```
class CustomOracleBoolean(TypeDecorator):
   """1, -1 trick"""
   
class CustomBoolean(TypeDecorator):
        if dialect == "oracle":
            return CustomOracleBoolean()
        else:
            return Boolean()

class Client(Base):
    __tablename__ = 'client'
    id = Column(Integer, primary_key=True)
    active = Column(CustomBoolean)  # 1,-1 trick for oracle, regular sqlalchemy boolean for other db vendors
```

I have not found a simple way to do that.
During experiments, I found a workaround to make `column_expression()` aware of current dialect (using `load_dialect_impl`), but it is more difficult for `comparator_factory`.

**Reply by CaselIT** [reply]
You can use with_variant in the types.

@zzzeek it seems very hard to find that functionality, since we have had other questions regarding it. Not sure if we could better document it.

@NicolasM-Forsk how did you search for that functionality (to understand if maybe some keyword are missing in the docs)

**Reply by NicolasM-Forsk** [reply]
sqlalchemy TypeDecorator dialect
sqlalchemy type per dialect
sqlalchemy customize type postgres
sqlalchemy same model different db vendors
comparator_factory per dialect
...

**Reply by NicolasM-Forsk** [reply]
When using `with_variant`, `comparator_factory` is no longer called.
Is there a workaround for this?

Fetch is not serialized when I compile the query:

Ex:
```
stmt = (
  select(Client.active).select_from(Client)
    .filter_by(active=True)
    .limit(200) ***
)
print(stmt.compile(engine))
SELECT CASE WHEN (client.active != 0) THEN 1 ELSE client.active END AS active 
FROM client 
WHERE CASE WHEN (client.active != 0) THEN 1 ELSE client.active END = 1
 LIMIT ? OFFSET ?
```

```
stmt = (
  select(Client.active).select_from(Client)
    .filter_by(active=True)
    .fetch(200) ***
)
print(stmt.compile(engine))
SELECT CASE WHEN (client.active != 0) THEN 1 ELSE client.active END AS active 
FROM client 
WHERE CASE WHEN (client.active != 0) THEN 1 ELSE client.active END = 1
```

**Reply by NicolasM-Forsk** [reply]
@CaselIT Do you know if with [this implementation](https://github.com/sqlalchemy/sqlalchemy/discussions/8219#discussioncomment-3100716) can work with `with_variant`? `comparator_factory` is not called through `Variant` type, there already have been discussions with per-case solution.

**Reply by CaselIT** [reply]
Not sure, have you tried doing the opposite variant? Meaning something like `CustomImpl().with_variant(Boolean, ...)`

@zzzeek thoughts?

**Reply by zzzeek** [reply]
> When using `with_variant`, `comparator_factory` is no longer called. Is there a workaround for this?

if you mean comparator_factory on the base type, this is [fixed in 2.0](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#with-variant-clones-the-original-typeengine-rather-than-changing-the-type).   there's no workaround in 1.4 for variant itself, you would instead stick with TypeDecorator and implement [load_dialect_impl](https://docs.sqlalchemy.org/en/14/core/custom_types.html?highlight=typedecorator#sqlalchemy.types.TypeDecorator.load_dialect_impl).

**Reply by NicolasM-Forsk** [reply]
> Not sure, have you tried doing the opposite variant? Meaning something like `CustomImpl().with_variant(Boolean, ...)`

`CustomImpl().with_variant(Boolean, ...)` : comparator is always applied (no matter the actual db vendor)
`Boolean().with_variant(CustomImpl, ...)` : comparator is never applied (no matter the actual db vendor)

I'd like to apply the comparator only for Oracle (ie for `CustomImpl`).

> there's no workaround in 1.4 for variant itself, you would instead stick with TypeDecorator and implement [load_dialect_impl](https://docs.sqlalchemy.org/en/14/core/custom_types.html?highlight=typedecorator#sqlalchemy.types.TypeDecorator.load_dialect_impl).

I don't see how to use `load_dialect_impl` to modify the behaviour of `comparator_factory`.
`comparator_factory` is called before `load_dialect_impl`.

**Reply by zzzeek** [reply]
> I don't see how to use `load_dialect_impl` to modify the behaviour of `comparator_factory`. `comparator_factory` is called before `load_dialect_impl`.

well you can't.  comparator factory is the in-Python behavior of your expressions and controls what kind of Python structure is generated, all of which is well before anything regarding an engine or dialect is in play.   if you wanted comparator factory to do something custom, and then the result of **that** would also be dialect specific, you would return a custom SQL construct, and then use [the compiles extension](https://docs.sqlalchemy.org/en/14/core/compiler.html) to give **that** object dialect-specific behaviors.  

I'm not understanding how this custom type is meant to work around a pattern in a specific Oracle database which has arbitrary integers that should be interpreted as boolean, but then you want that same type to work on non-Oracle databases which...have the same numerical pattern?  or are normal 0/1 values?     there's still a suggestion here that the overall problem space might be approached differently.

**Reply by NicolasM-Forsk** [reply]
Yes, other databases use regular 0/1 values or boolean values, this is why I try to customize only Oracle behavior.
I think that we'll have to fix our dataset before supporting Oracle.

Compiler extension may be fine if I can pass comparison operator through, I'll give it a try.
Thanks for your help

**Reply by zzzeek** [reply]
so that would then indicate that these non 0/1 values are not actually meaningful.  why not just migrate the Oracle databases to have 0/1 like the other backends?

**Reply by NicolasM-Forsk** [reply]
Yeah I would like to do so, but we cannot deploy database migrations often, and we have to handle retro compatibility.
Reading SQLAlchemy documentation, I thought it would be possible to use custom types as a temporary workaround.

**Reply by zzzeek** [reply]
it is but this is a really wrangly one because you are trying to work around entirely different data on just one backend

**Reply by CaselIT** [reply]
another alternative may be to use a computed column in oracle to handle the difference. So in sqlalchemy you only have to care about boolean

**Reply by NicolasM-Forsk** [reply]
Custom function works fine:
```
@compiles(normalized_bool, 'default')
def normalized_bool_default(element, compiler, **kw):
    arg = list(element.clauses)[0]
    return compiler.process(arg, **kw)

@compiles(normalized_bool, 'oracle')
def normalized_bool_oracle(element, compiler, **kw):
    arg = list(element.clauses)[0]
    res1 = case(
        (type_coerce(arg, Integer()) != literal_column("0", type_=Integer), literal_column("1", type_=Integer)),
        else_=type_coerce(arg, Integer())
    )
    res2 = type_coerce(res1, Boolean())
    return compiler.process(res2, **kw)

class CustomOracleBoolean(TypeDecorator):
    def column_expression(self, col):
       ...
       return normalized_bool(expr)

    class comparator_factory(Boolean.Comparator):
        ...
        op1 = normalized_bool(clause)
        op2 = normalized_bool(clause)
        return op(op1, op2, **kwargs)
```

Do you know if such transformation can be applied when applying `group by` and aggregation functions to a column?

**Reply by NicolasM-Forsk** [reply]
> another alternative may be to use a computed column in oracle to handle the difference. So in sqlalchemy you only have to care about boolean

I cannot because of product constraints

**Reply by zzzeek** [reply]
It seems likely that group_by / order_by don't apply any functions from the TypeEngine, since column_expression() , result_processor() etc. are all about the interaction between the SQL datatype and the Python value.   so you will need to apply your normalized_bool() construct explicitly.   it's possible to use more event hooks like do_orm_execute() or `@compiles` again to get at this as well within SELECT statements, it might require a bit of plumbing (like cloning the statement first) to do it correctly.
